gfx.h

```diff
@@ -3344,54 +3345,80 @@ class GfxInternal
             if(buffer_offset >= src.size)
                 break;  // further mips aren't available
             uint64_t const buffer_row_pitch = row_sizes[mip_level];
-            uint64_t const texture_row_pitch = subresource_footprints[mip_level].Footprint.RowPitch;
             uint64_t const buffer_size = (uint64_t)num_rows[mip_level] * buffer_row_pitch;
             if(buffer_offset + buffer_size > src.size)
                 return GFX_SET_ERROR(kGfxResult_InvalidOperation, "Cannot copy to mip level %u from buffer object with insufficient storage", mip_level);
-            if(buffer_row_pitch != texture_row_pitch || // we must respect the 256-byte pitch alignment
-               GFX_ALIGN(buffer_offset, D3D12_TEXTURE_DATA_PLACEMENT_ALIGNMENT) != buffer_offset)
-            {
-                uint64_t texture_upload_buffer_size = num_rows[mip_level] * static_cast<uint64_t>(subresource_footprints[mip_level].Footprint.RowPitch);
-                if(texture_upload_buffer_.size < texture_upload_buffer_size)
-                {
-                    destroyBuffer(texture_upload_buffer_);
-                    texture_upload_buffer_size += (texture_upload_buffer_size + 2) >> 1;
-                    texture_upload_buffer_size = GFX_ALIGN(texture_upload_buffer_size, 65536);
-                    texture_upload_buffer_ = createBuffer(texture_upload_buffer_size, nullptr, kGfxCpuAccess_None);
-                    if(!texture_upload_buffer_)
-                        return GFX_SET_ERROR(kGfxResult_OutOfMemory, "Unable to allocate memory to upload texture data");
-                    texture_upload_buffer_.setName("gfx_TextureUploadBuffer");
-                }
-                texture_upload_buffer = &buffers_[texture_upload_buffer_];
-                SetObjectName(*texture_upload_buffer, texture_upload_buffer_.name);
-                Buffer &gfx_buffer = buffers_[src]; SetObjectName(gfx_buffer, src.name);
-                if(src.cpu_access == kGfxCpuAccess_None) transitionResource(gfx_buffer, D3D12_RESOURCE_STATE_COPY_SOURCE);
-                transitionResource(*texture_upload_buffer, D3D12_RESOURCE_STATE_COPY_DEST);
-                submitPipelineBarriers();   // transition our resources if needed
-                for(uint32_t i = 0; i < num_rows[mip_level]; ++i)
-                    command_list_->CopyBufferRegion(texture_upload_buffer->resource_, i * texture_row_pitch, gfx_buffer.resource_, i * buffer_row_pitch + buffer_offset, buffer_row_pitch);
-            }
-            Buffer &gfx_buffer = buffers_[src]; SetObjectName(gfx_buffer, src.name);
-            if(texture_upload_buffer != nullptr) transitionResource(*texture_upload_buffer, D3D12_RESOURCE_STATE_COPY_SOURCE);
-            else if(src.cpu_access == kGfxCpuAccess_None) transitionResource(gfx_buffer, D3D12_RESOURCE_STATE_COPY_SOURCE);
-            transitionResource(gfx_texture, D3D12_RESOURCE_STATE_COPY_DEST);
-            submitPipelineBarriers();   // transition our resources if needed
```

Comment:
I also noticed that the existing buffer to texture functions does a series of smaller CopyBufferRegion calls when buffer and texture pitch dont match. However I found that by just setting the Footprint.RowPitch in the CopyTextureRegion function makes this unnecessary. I checked this in PIX using input buffers containing mip chains that trigger this code path and it works identically after this change. It is also still round trip identical when combined with the textureToBuffer function.
This change is in its own commit so can be skipped/reverted if not wanted.